### PR TITLE
Abstracting source relations for enhanced covering index rewriting

### PR DESCRIPTION
**Pull request by dai-chen**
### Description

This PR introduces new abstractions for the covering index query rewriter, facilitating support for different source table relation matching and rewriting. This enhancement paves the way for future support of Iceberg table relations.

#### PR Planned

- [x] https://github.com/opensearch-project/opensearch-spark/pull/318
- [x] https://github.com/opensearch-project/opensearch-spark/pull/325 *[Track in Iceberg project separately]*
- [ ] https://github.com/opensearch-project/opensearch-spark/pull/391 *[Current PR]*
- [x] Add logging or whyNot API to explain why/whyNot index applied *[Logging added in this PR]*
- [ ] Support SQL hint for Spark conf, ex. CV rewrite, hybrid scan *[TBD]*
- [ ] Support partial covering index rewrite *[TBD]*

#### Changes

Added new `FlintSparkSourceRelationProvider` and `FlintSparkSourceRelation`  abstraction. Please see Scala doc for its responsibility in details. Basically,

- `FlintSparkSourceRelationProvider`: determine if a given logical relation can be supported by Flint optimizer.
- `FlintSparkSourceRelation`: provide all information required by query rewriting for a specific source relation.

Will refactor `ApplyFlintSparkSkippingIndex` and `FlintSparkValidationHelper.isTableProviderSupported` based on these in future.

Screenshot 2024-04-30 at 11 55 56 AM

### Testing

<pre>
spark-sql> CREATE INDEX all ON myglue.ds_tables.http_logs
         > (
         >   `@timestamp`,
         >   clientip,
         >   request,
         >   status,
         >   size
         > );

scala> sc.setLogLevel("INFO")
scala> sql("EXPLAIN SELECT clientip FROM myglue.ds_tables.http_logs WHERE status != 200").show

# Logging explains whether and why the index is applied
24/05/03 17:51:17 INFO FlintSparkSourceRelationProvider: Loaded source relation providers [file]
24/05/03 17:51:17 INFO ApplyFlintSparkCoveringIndex: Provider [file] can match sub plan LogicalRelation
24/05/03 17:51:18 INFO ApplyFlintSparkCoveringIndex: Found covering index 
[flint_myglue_ds_tables_http_logs_all_index] on table myglue.ds_tables.http_logs
24/05/03 17:51:18 INFO ApplyFlintSparkCoveringIndex:
 Is covering index flint_myglue_ds_tables_http_logs_all_index applicable: true
   Index state: Some(active)
   Index filter condition: None
   Columns required: Set(clientip, status)
   Columns indexed: Set(@timestamp, request, size, clientip, status)
</pre>

### Issues Resolved

https://github.com/opensearch-project/opensearch-spark/issues/298

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/source/FlintSparkSourceRelation.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * source data formats (like Spark built-in File, Delta table, Iceberg, etc.), hiding the
+ * specifics of each data source implementation.
+ */
+trait FlintSparkSourceRelation {
+
+  /**
+   * @return
+   *   the concrete logical plan of the relation associated
+   */
+  def plan: LogicalPlan
```
Comment (penghuo) [inline flint-spark-integration/src/main/scala/org/opensearch/flint/spark/source/FlintSparkSourceRelation.scala:22]: does it represent DataSourceV2Relation / BaseRelation? or the whole execution plan?

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/source/FlintSparkSourceRelationProvider.scala [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+   * @param spark
+   *   the Spark session
+   * @return
+   *   a sequence of source relation provider
+   */
+  def getAllProviders(spark: SparkSession): Seq[FlintSparkSourceRelationProvider] = {
+    var relations = Seq[FlintSparkSourceRelationProvider]()
+
+    // File source is built-in supported
+    relations = relations :+ new FileSourceRelationProvider
```
Comment (penghuo) [inline flint-spark-integration/src/main/scala/org/opensearch/flint/spark/source/FlintSparkSourceRelationProvider.scala:67]: how to support other providers?

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/source/file/FileSourceRelation.scala [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import org.apache.spark.sql.catalyst.expressions.AttributeReference
+import org.apache.spark.sql.execution.datasources.LogicalRelation
+
+/**
+ * Concrete source relation implementation for Spark built-in file-based data sources.
+ *
+ * @param plan
+ *   the `LogicalRelation` that represents the plan associated with the File-based table
+ */
+case class FileSourceRelation(override val plan: LogicalRelation)
```
Comment (penghuo) [inline flint-spark-integration/src/main/scala/org/opensearch/flint/spark/source/file/FileSourceRelation.scala:19]: DataSourceV2Relation also covered?

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/source/FlintSparkSourceRelationProvider.scala (hunk 2 above)
Comment (dai-chen) [reply flint-spark-integration/src/main/scala/org/opensearch/flint/spark/source/FlintSparkSourceRelationProvider.scala:67]: Here is an example from my previous draft PR: https://github.com/opensearch-project/opensearch-spark/pull/325/files#diff-1b438c3d2fc10574cefff7b14912704579f2d70b87c1aa7d03bd70dd9f80f522R70-R75.

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/source/FlintSparkSourceRelation.scala (hunk 1 above)
Comment (dai-chen) [reply flint-spark-integration/src/main/scala/org/opensearch/flint/spark/source/FlintSparkSourceRelation.scala:22]: yes, it represents the LogicalPlan node (`LogicalRelation` or `DataSourceV2Relation`)

**Review by penghuo — APPROVED** [review]

**Pull request merged**